gateway/gateway.go

```diff
@@ -0,0 +1,366 @@
+package gateway
+
+import (
+	"context"
+	"fmt"
+	"html/template"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/gin-contrib/location"
+	"github.com/gin-contrib/static"
+	"github.com/gin-gonic/gin"
+	logger "github.com/ipfs/go-log"
+	assets "github.com/jessevdk/go-assets"
+	"github.com/rs/cors"
+	gincors "github.com/rs/cors/wrapper/gin"
+	"github.com/textileio/fil-tools/index/ask"
+	"github.com/textileio/fil-tools/index/miner"
+	"github.com/textileio/fil-tools/index/slashing"
+	"github.com/textileio/fil-tools/reputation"
+)
+
+var log = logger.Logger("gateway")
+
+// fileSystem extends the binary asset file system with Exists,
+// enabling its use with the static middleware.
+type fileSystem struct {
+	*assets.FileSystem
+}
+
+// Exists returns whether or not the path exists in the binary assets.
+func (f *fileSystem) Exists(prefix, path string) bool {
+	pth := strings.TrimPrefix(path, prefix)
+	if pth == "/" {
+		return false
+	}
+	_, ok := f.Files[pth]
+	return ok
+}
+
+// Gateway provides HTTP-based access to Textile.
+type Gateway struct {
+	addr             string
+	server           *http.Server
+	askIndex         *ask.AskIndex
+	minerIndex       *miner.MinerIndex
+	slashingIndex    *slashing.SlashingIndex
+	reputationModule *reputation.Module
+}
+
+// NewGateway returns a new gateway.
+func NewGateway(
+	addr string,
+	askIndex *ask.AskIndex,
+	minerIndex *miner.MinerIndex,
+	slashingIndex *slashing.SlashingIndex,
+	reputationModule *reputation.Module,
+) *Gateway {
+	return &Gateway{
+		addr:             addr,
+		askIndex:         askIndex,
+		minerIndex:       minerIndex,
+		slashingIndex:    slashingIndex,
+		reputationModule: reputationModule,
+	}
+}
+
+// Start the gateway.
+func (g *Gateway) Start() {
+	gin.SetMode(gin.ReleaseMode)
+	router := gin.Default()
+	router.Use(location.Default())
+
+	// @todo: Config based headers
+	options := cors.Options{}
+	router.Use(gincors.New(options))
+
+	temp, err := loadTemplate()
+	if err != nil {
+		log.Fatal(err)
+	}
+	router.SetHTMLTemplate(temp)
+
+	router.Use(static.Serve("", &fileSystem{Assets}))
+
+	router.GET("/asks", g.asksHandler)
+	router.GET("/miners", g.minersHandler)
+	router.GET("/slashing", g.slashingHandler)
+	router.GET("/reputation", g.reputationHandler)
+
+	router.GET("/", func(c *gin.Context) {
+		c.Request.URL.Path = "/asks"
+		router.HandleContext(c)
+	})
+
+	router.NoRoute(func(c *gin.Context) {
+		g.render404(c)
+	})
+
+	g.server = &http.Server{
+		Addr:    g.addr,
+		Handler: router,
+	}
+
+	errc := make(chan error)
+	go func() {
+		errc <- g.server.ListenAndServe()
+		close(errc)
+	}()
+	go func() {
+		for {
+			select {
+			case err, ok := <-errc:
+				if err != nil {
+					if err == http.ErrServerClosed {
+						return
+					}
+					log.Errorf("gateway error: %s", err)
```

Comment:
Should a `return` be added after this line? I think that even if that isn't the case, the next for loop will enter in the `if !ok` and exit indirectly, but that sounds like a lucky consequence than on purpose?

Reply:
Yea I think you're right, added the `return`.